File: server/services/image_analysis_agent.py
"""
Image Analysis Agent - VLM-powered worksheet evaluation and image analysis
"""
import json
import base64
from typing import Dict, List, Any
from shared_config import (
    model, is_valid_api_key, logger, parse_gemini_response
)

async def analyze_image(image_data: str) -> Dict[str, Any]:
    """Analyze image using Gemini Vision"""
    if not model:
        return {
            "description": "Image analysis requires valid Gemini API key",
            "elements": [],
            "educational_value": "Cannot analyze without API access"
        }
    
    try:
        # Decode base64 image if provided
        if image_data.startswith('data:image'):
            header, data = image_data.split(',', 1)
            image_bytes = base64.b64decode(data)
        else:
            image_bytes = base64.b64decode(image_data)
        
        prompt = """Analyze this image for educational purposes. Provide a comprehensive analysis suitable for Indian educational context.

**Requirements:**
1. Identify all visible objects, people, or scenes
2. Suggest educational concepts that could be taught
3. Provide age-appropriate questions for different grade levels
4. Include cultural context relevant to Indian education
5. Suggest hands-on activities based on the image

**Output Format (JSON):**
{
  "description": "detailed description of the image",
  "elements": ["list of key elements visible"],
  "educationalConcepts": ["concepts that can be taught"],
  "gradeLevel": "suggested grade levels for this content",
  "subjectAreas": ["subjects this image relates to"],
  "discussionQuestions": {
    "beginner": ["questions for younger students"],
    "intermediate": ["questions for middle grade students"],
    "advanced": ["questions for older students"]
  },
  "activities": ["hands-on activities inspired by this image"],
  "culturalContext": "relevance to Indian culture/education",
  "vocabularyWords": ["new words students can learn"],
  "realWorldConnections": ["how this relates to students' daily lives"],
  "assessmentIdeas": ["ways to assess student understanding"],
  "materialsSuggested": ["materials needed for related activities"]
}

Provide rich educational insights that teachers can immediately use!"""

        # Create image part for Gemini
        image_part = {
            "mime_type": "image/jpeg",  # Assume JPEG, adjust as needed
            "data": image_bytes
        }
        
        response = model.generate_content([prompt, image_part])
        
        result = parse_gemini_response(response.text)
        
        # Add metadata
        result["aiGenerated"] = True
        result["model"] = "gemini-1.5-pro-latest"
        result["analysisType"] = "educational_image_analysis"
        
        logger.info("Successfully analyzed image for educational content")
        return result
        
    except Exception as error:
        logger.error(f"Gemini Vision API Error: {error}")
        return {
            "description": "Image analysis failed",
            "elements": [],
            "educational_value": f"Analysis failed due to error: {error}",
            "note": f"Image analysis error: {error}"
        }

File: server/services/image-analysis-agent.ts
import { model, isValidApiKey, logger, parseGeminiResponse } from './shared-config';

function generateMockImageAnalysis(imageData: string): any {
  /**Generate mock image analysis when Gemini is not available*/
  return {
    description: "Sample image analysis for educational content",
    educationalContent: ["Sample educational point 1", "Sample educational point 2"],
    suggestedActivities: ["Discussion about the image", "Drawing activity"],
    vocabulary: ["Sample word 1", "Sample word 2"],
    note: "Demo content - Gemini service unavailable"
  };
}

export async function analyzeImage(imageData: string): Promise<any> {
  /**Analyze image for educational content using Image Analysis Agent*/
  try {
    if (!isValidApiKey || !model) {
      logger.warn("Gemini API not configured, using mock content");
      return generateMockImageAnalysis(imageData);
    }

    // For now, we'll use text-based analysis since image analysis requires
    // proper image data handling. In a real implementation, you'd need to
    // handle base64 images or file uploads properly.
    
    const prompt = `Analyze the following image data for educational content.
    
    Image data: ${imageData.substring(0, 100)}...
    
    Requirements:
    - Identify educational elements in the image
    - Suggest learning activities based on the image
    - Extract vocabulary words
    - Provide cultural context if relevant
    - Make it suitable for Indian classrooms
    
    Return the response as a JSON object with the following structure:
    {
      "description": "Detailed description of the image",
      "educationalContent": ["educational point 1", "educational point 2"],
      "suggestedActivities": ["activity 1", "activity 2"],
      "vocabulary": ["word 1", "word 2"],
      "culturalContext": "Cultural relevance for Indian students",
      "gradeLevel": "suitable grade levels",
      "subjectAreas": ["subject 1", "subject 2"]
    }`;

    const result = await model.generateContent(prompt);
    const response = await result.response;
    const text = response.text();
    
    const parsedResponse = parseGeminiResponse(text);
    
    if (parsedResponse.error) {
      logger.error(`Failed to parse Gemini response: ${parsedResponse.error}`);
      return generateMockImageAnalysis(imageData);
    }
    
    return {
      ...parsedResponse,
      culturalContext: parsedResponse.culturalContext || "Culturally relevant for Indian students"
    };
    
  } catch (error) {
    logger.error(`Image Analysis Agent Error: ${error}`);
    return generateMockImageAnalysis(imageData);
  }
} 